### ISSUE TITLE
RMA win based buffer pre-registration

### ISSUE THREAD

**Issue by minsii** [pmodels/mpich#3585]
The RMA `window creation interface` can be used with `info hints` to pre-register special communication buffers (e.g.,send/recv buffer, collective buffer, RMA local buffer), in order to reduce the overhead of performance-critical issuing operation. Specific optimizations include:
- SHM-aware communication
- Direct RDMA (CPU memory) without per-message network memory registration
- Direct GPU DMA/RDMA data transfer without per-message GPU memory pinning overhead

**Comment by pavanbalaji** [comment]
For networks and GPUs, I think `MPI_Alloc_mem` is sufficient.  The benefit of RMA windows is only for cases that need collective buffer allocation (e.g., shared memory).

**Comment by minsii** [comment]
You are right that we can preregister the local buffer and cache any local memory registration info through `MPI_Alloc_mem`. Can we optimize some cases through the window API if the later communication requires remote memory regisration info ? This info can be pre-exchanged at window creation time.
